File: Forge/src/main/java/hantonik/atomic/core/registration/DelayedSupplier.java
package hantonik.atomic.core.registration;

import lombok.Setter;
import org.jetbrains.annotations.Nullable;

import java.util.function.Supplier;

public class DelayedSupplier<T> implements Supplier<T> {
    @Setter
    @Nullable
    private Supplier<T> supplier;

    @Override
    public T get() {
        if (this.supplier == null)
            throw new IllegalStateException("Attempted to call DelayedSupplier::get() before the supplier was set.");

        return this.supplier.get();
    }
}

File: Forge/src/main/kotlin/hantonik/atomic/core/registration/DelayedSupplier.kt
package hantonik.atomic.core.registration

class DelayedSupplier<T> : () -> T {
    var supplier: (() -> T)? = null

    override fun invoke(): T {
        this.supplier ?: throw IllegalStateException("Attempted to call DelayedSupplier::invoke() before the supplier was set.")

        return this.supplier!!.invoke()
    }
}